File: java-vfs/src/main/java/emg/java/vfs/commands/Ls.java
package emg.java.vfs.commands;

import emg.java.vfs.files.DirEntry;
import emg.java.vfs.filesystem.State;

import java.util.List;

import static emg.java.vfs.extensions.ListExtensions.head;
import static emg.java.vfs.extensions.ListExtensions.tail;

public class Ls extends Command {
    @Override
    public State apply(State state) {
        var contents = state.wd().contents();
        var niceOutput = createNiceOutput(contents);
        return state.setMessage(niceOutput);
    }

    private String createNiceOutput(List<DirEntry> contents) {
        if (contents.isEmpty()) {
            return "";
        } else {
            var entry = head(contents);
            return String.format("%s [%s]%n%s", entry.name(), entry.getType(), createNiceOutput(tail(contents)));
        }
    }
}

File: kotlin-vfs/src/main/kotlin/emg/kotlin/vfs/commands/LS.kt
package emg.kotlin.vfs.commands

import emg.kotlin.vfs.extensions.head
import emg.kotlin.vfs.extensions.tail
import emg.kotlin.vfs.files.DirEntry
import emg.kotlin.vfs.filesystem.State

class LS : Command() {
    private fun createNiceOutput(contents: List<DirEntry>): String {
        return if (contents.isEmpty()) ""
        else {
            val entry = contents.head()
            entry.name + " [${entry.getType()}]\n${createNiceOutput(contents.tail())}"
        }
    }

    override fun apply(state: State): State {
        val contents = state.wd.contents
        val niceOutput = createNiceOutput(contents)
        return state.setMessage(niceOutput)
    }
}